[FLINK-7832] [flip6] Extend SlotManager to report free slots per TM

tillrohrmann opened the pull request:
## What is the purpose of the change

Extend the `SlotManager` such that we count the free slots per `TaskManager`. This has the advantage that we don't have to iterate over all registered slots and aggregate their state in order to decide whether a TaskManager is idle or not. Moreover, it allows to easily query how many free slots every `TaskManager` still has.

## Brief change log

- Fail if slot belongs to a unregistered TaskManager
- Add more sanity checks
- Make the TaskManagerSlot state transitions clearer
- Introduce proper TaskManagerSlot state enum and state transitions
- Refactor SlotManager for better maintainability
- Add free slot counting

## Verifying this change

This change is already covered by existing tests, such as `SlotManagerTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/TaskManagerSlot.java:
```diff
@@ -74,22 +83,51 @@ public AllocationID getAllocationId() {
+		assignedSlotRequest = null;
+	}
+
+	public void updateAllocation(AllocationID allocationId) {
```
Can you explain the difference between updateAllocation and completeAllocation in terms of when they are used? (I would've expected that the slot lifecycle is along the lines of FREE <-> PENDING -> ALLOCATED -> FREE), but this doesn't appear to be the case as a free slot can be allocated without ever being in a pending state).

tillrohrmann replied:
`updateAllocation` is used to model the state transition from `FREE` to `ALLOCATED`. This state transition is necessary, because the ground truth is hold by the `TaskExecutor`. Thus, it can be the case that the `ResourceManager` receives a `SlotReport` where it says that slot `x` is allocated with this `allocationId`. This can be the case, for example, in the RM failover case.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/TaskManagerSlot.java:
```diff
@@ -74,22 +83,51 @@ public AllocationID getAllocationId() {
+		Preconditions.checkState(Objects.equals(allocationId, assignedSlotRequest.getAllocationId()), "Mismatch between allocation id of the pending slot request.");
+
+		state = State.ALLOCATED;
+		this.allocationId = Preconditions.checkNotNull(allocationId);
```
I think `Objects.equals` takes care of this.

tillrohrmann replied:
But will move it to the top to make it clearer.

zentol replied:
this wasn't about correctness but clarity, so it's good that you changed it :)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -418,26 +415,17 @@ public void freeSlot(SlotID slotId, AllocationID allocationId) {
-							taskManagerRegistration.markIdle();
-						}
+					if (taskManagerRegistration == null) {
+						throw new IllegalStateException("Trying to free a slot from a TaskManager " +
```
Yes, will add it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -556,14 +537,30 @@ private void registerSlot(
+
+				return true;
+			} else {
+				throw new IllegalStateException("Trying to update a slot from a TaskManager " +
```
Would say this is not necessary.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -556,14 +537,30 @@ private void registerSlot(
+		}
+	}
+
+	private void updateSlotInternal(TaskManagerSlot slot, TaskManagerRegistration taskManagerRegistration, @Nullable AllocationID allocationId) {
```
True. The intention was to have the lookup logic in `updateSlot` because this is not needed for `freeSlot` and `removeSlotRequest`. Will rename `updateSlotInternal` into `updateSlotState`.

tillrohrmann commented:
Thanks for your review @zentol. I've addressed your comments and hope to have clarified the `updateAllocation` call for the state transition `FREE` -> `ALLOCATED`. 

If there are not other objections, then I would like to merge this PR once Travis gives green light.
